### PR TITLE
Optimize FEPointEvaluation::integrate() with tensor product path

### LINKED ISSUE #11865: Improve FEPointEvaluation::integrate()
This is a reminder that `FEPointEvaluation::integrate()` still used the slow code path in contrast to `FEPointEvaluation::evaluate()`.

### PR DESCRIPTION
**Pull request by kronbichler**
Fixes #11865. This is a bit late for the release, so I would certainly not insist in getting it in. The important discussion we should have before the release is about the interface we have in `MappingQGeneric::transform_variable` because we should not change that again.

### CODE REVIEW

File: tests/fe/point_evaluation_04.output
Comment (kronbichler) [inline tests/fe/point_evaluation_04.output:36]: This is one thing we should discuss: The optimized path only writes into those DoFs where the selected component is non-zero, leaving the rest unchanged. The old implementation did zero the other parts. What do we want to have? I would find it more natural to only touch those that belong to the selected part, but it would one more round of checks with `FEValues`. Or I just add a zeroing operation to the tensor product path.

**Review by peterrum — APPROVED** [review]
Looks good to me. Thanks for tackling the issue!

Please go ahead and create a PR for changing the interface of `MappingQGeneric` so that we can integrate that before the release!

File: tests/fe/point_evaluation_04.output
Comment (kronbichler) [reply tests/fe/point_evaluation_04.output]: I now pushed a new version where the behavior is clearly documented (we clear `solution_values` in that function, consistent with #12235). The new code is hence consistent with all previous tests, showing that the new functionality is correct.

**Comment by kronbichler** [comment]
@bergbauer would you mind having a look as well to see if you agree with this change? Having another pair of eyes looking at things will certainly not hurt, now that the release is close.

**Review by peterrum — APPROVED** [review]
If there are no objection, I will merge this PR before creating the 9.3 branch later today.

**Review by bergbauer — APPROVED** [review]
Looks good to me. Cannot wait to test the new functionality!

**Comment by peterrum** [comment]
@kronbichler Have you also run `ctest -j30  -R "remote_point_evaluation/"`?  There are some more involved tests using `FEPointEvaluation`.

**Comment by kronbichler** [comment]
Yes, I ran all tests containing `point_evaluation`, which includes `remote_point_evaluation`, and they pass. Here are the last few lines of my output:
```
47/48 Test #10551: remote_point_evaluation/vector_tools_evaluate_at_points_01.mpirun=8.release ...   Passed   13.24 sec
48/48 Test #10553: remote_point_evaluation/vector_tools_evaluate_at_points_02.mpirun=8.release ...   Passed   17.83 sec

100% tests passed, 0 tests failed out of 48

Total Test time (real) = 198.33 sec
```

**Comment by peterrum** [comment]
Thanks!

**Comment by kronbichler** [comment]
Once we have merged this and possibly #12221, I will also run the test suite on `node29`, https://cdash.43-1.org/viewSite.php?siteid=3&project=1&currenttime=1621299600, to get a picture on the state.

**Comment by peterrum** [comment]
CI is happy. GitHub actions seem to be down again; sigh... I am bit reluctant to merge (due to our bad experience with not running the test in the last few days); nevertheless it would be good to proceed here, since we have one follow-up PR with some major interface changes which should be part of the release. Any opinions?

**Comment by kronbichler** [comment]
At least each of Windows, OSX and Linux each had one system which ran successfully through all tests.

**Comment by peterrum** [comment]
Merging to be able to proceed with the follow-up PRs.

**Comment by kronbichler** [comment]
It looks like this PR makes `point_evaluation_03.release` and `point_evaluation_09.release` fail on node29 https://cdash.43-1.org/test/6495522, but only with AVX-512 and release mode and only for the vector-valued cases. It almost looks like a compiler bug (present both for gcc-9 and gcc-10). I'll investigate.

**Comment by kronbichler** [comment]
I cannot see this on current master any more, it seems #12236 (likely) or #12245 did indeed eliminate whatever problem we might have had. Let us wait for node29 to come up with new results. I will keep an eye on the tests to ensure things remain stable with AVX-512.

**Pull request merged**
